### PR TITLE
AoPS (Art of Problem Solving) forum crawler

### PR DESCRIPTION
**Pull request by TheSil**
Made a simple AoS crawler, it works by specifying either single post, or time range (currently relative dates, not absolutes). Also the AoS forums do not work on same principle as the MathSE, thats why I use time (number of days) rather than number of pages. Also I had to use ajax php requests since that is the way the data are fetched.

Anyway few examples to try:
-c 6 -p 1639347
This fetches https://artofproblemsolving.com/community/c6h1639347_sum_of_squares .

-c 7 -n 0 -o 7
This fetches posts from category 7 (https://artofproblemsolving.com/community/c7), posts that are 7 days old

-c 7 -n 7 -o 14
This fetches posts from category 7 (https://artofproblemsolving.com/community/c7), posts that are 7-14 days old

**Few considerations:**
The relative time API might not be best, could be perhaps improved by using absolute times, could improve this in the future. I see biggest challenge with organizing these different forums categories, but it seems you can just fetch following threads to get vas majority of posts on AoS:
-c 3 (Middle School Math, > 33k posts)
-c 4 (High School Math > 71k posts)
-c 5 (Contests & Programs > 17k posts)
-c 6 (High School Olympiads > 214k posts)
-c 7 (College Math > 78k posts)

Also the folders and files layout, I have added prefix aos to files in order todistinguish between SE and Aos posts, you might want to adjust that.

I am using few more libs though, slimit for javascript parsing and certifi since on my machine the SSL links did not work (maybe Windows only issue, I dont know). Also along the way I encountered few issues witn encoding and parsing even the Math.SE posts, so I fixed those, but feel free to ommit/remove it, but I believe it improved the parser. 

Although it works fine, I imagine there might be some additional updates, such as putting common crawlers stuff into common file, but I mainly botherred with functionality at this point, to get this going.

### CODE REVIEW

**Comment by TheSil** [comment]
I wonder why this Travis thing failed, but It seems like it is unrelated to the requesting python changes (C compilation issues).

**Comment by t-k-** [comment]
Thank you!

The Travis failing is not your issue, it is fixed in my research branch (24f9afa9b8e09f6b55b582ab261f27a95fc6d3a4).

**Comment by TheSil** [comment]
Good to hear, hope it works for you! How are the crawlers scheduled by the way? I mean how often do you fetch the data?

Also a side note, the bbcode stuff etc It was mainly because the html previews looked shitty when I opened them in chrome (it marked every instance of it red, which is quite a lot on some pages...). Also I believe the certifi caused issues with math SE.

**Comment by t-k-** [comment]
Yes, it seems working perfectly. Again, I am very thankful you send me the request.
I was just trying to bring the patches from my research branch to master (there are some important updates for crawler scripts, e.g., Math Stack Exchange recently becomes very sensitive to crawling and very often prevent our scripts from crawling if we keep a constant frequency, see: 6b9cf283b44b8f8cd090e611156a0716f2d76bbe), sorry for the pause to get back.

For your question,
It is a little embarrassing that I do not have a script for updating the index regularly, and the online version of Approach0 is not updated for months! (Although at my lab I am running a crawler and I have basically an updated corpus of Math Stack Exchange, but did not get a chance to update the indices of online demo)
In fact, I am recently focus on improving the model and have achieved a significant accuracy improvement (will publish a paper soon hopefully). I feel I do not have enough time to write an "index update scheduler" because that is not my current focus, A0 is more like a demo instead of a serious search engine now.

But since you make it possible to have a new site indexed now, I will take some time to expand the index of approach0.xyz, to let AoPS in, and see how it looks like.

**Comment by t-k-** [comment]
@TheSil Thank you for your reply. I looked into their JSON data, indeed, they use LaTeX macro to represent some commonly seen symbols such as "=", in `post_canonical` field:

However, in `post_rendered` field, they use img with `alt` attribute that is written in commonly seen LaTeX.

Here is the first post I extracted from c6h246421:
```javascript
{
  "post_id": 1353555,
  "topic_id": 246421,
  "poster_id": 22793,
  "post_rendered": "Show that the equation <img src=\"\/\/latex.artofproblemsolving.com\/8\/3\/3\/833d8b15a3a9e9290c38c87a6d31418bf8a39d84.png\" class=\"latex\" alt=\"$ y^{37} = x^3 +11\\pmod p$\" style=\"vertical-align: -4px\" width=\"176\" height=\"19\" > is solvable for every prime <span style=\"white-space:nowrap;\"><img src=\"\/\/latex.artofproblemsolving.com\/6\/5\/f\/65f279776d6c37b2790a28c7a1616a695c9e1170.png\" class=\"latex\" alt=\"$ p$\" style=\"vertical-align: -3px\" width=\"10\" height=\"11\" >,<\/span> where <span style=\"white-space:nowrap;\"><img src=\"\/\/latex.artofproblemsolving.com\/8\/1\/4\/814f68625d1236215d50b95768d742ec64793489.png\" class=\"latex\" alt=\"$ p\\le 100$\" style=\"vertical-align: -3px\" width=\"61\" height=\"16\" >.<\/span>",
  "post_canonical": "Show that the equation $ y^{37} \\equal{} x^3 \\plus{}11\\pmod p$ is solvable for every prime $ p$, where $ p\\le 100$.",
  "username": "April",
  "reported": false,
  "is_thanked": false,
  "is_nothanked": false,
  "attachment": false,
  "thanks_received": 0,
  "nothanks_received": 0,
  "thankers": null,
  "deleted": false,
  "post_number": 1,
  "post_time": 1229904603,
  "num_edits": 0,
  "post_format": "bbcode",
  "last_edit_time": 0,
  "last_editor_username": "",
  "last_edit_reason": "",
  "admin": false,
  "avatar": "\/\/avatar.artofproblemsolving.com\/avatar_22793.jpg",
  "num_posts": 1270,
  "editable": false,
  "deletable": false,
  "show_from_start": true,
  "show_from_end": true
}
```
To extract `post_rendered` is not a good idea (hard to convert, and there might be other representations other than LaTeX). So it is reasonable to add back your rules to replace those `\equal{}` macro. So where do you get the list of such macro that we need to replace? Are you based on purely experience? Or we can find their rules somewhere (e.g.  in AoPS's site Javascript)?

**Comment by t-k-** [comment]
Committed. See 55a0a5adee299bcd3e6366c03dbdf6c9ff6c61c7

**Comment by TheSil** [comment]
Basically found those out just based on several pages inspected, those committed are all I encountered so far. I haven't found anything in the .js files (doesn't mean there isn't anythign though).

By the way today I noticed that actually only 30 posts of the topic get loaded on the first request, then subsequent Ajax requests with action a=fetch_posts_for_topic are required. Those 30 are enough most of the time, or maybe even all of the time for searching purposes, but nevertheless I am currently working on update that would be able to grab all posts in a topic. Ideally it could be possible to avoid the first request and fire the ajax.php instead already, thus parsing only JSON and not the classic forum topic html. I will see how it will go...

**Comment by TheSil** [comment]
Ok now loading all posts in a topic, see 959a13302581398d12d62daa195bc617b5e86252 . Unfortunately no speed up since the first request is required because of question title and to establish session. I don't know if you can pull  the code if I need to invoke another pull request to get few more updates in. I am merging your updates periodically, so the code should be compatible (including those things you reverted).

**Comment by t-k-** [comment]
@TheSil  Thank you for all the great work. I can merge your branch locally and push to Approach0 master, but if you want to make pull request, I definitely welcome.

**Comment by t-k-** [comment]
@TheSil  I have merged your changes, you do not need to make pull request now.

So topic in AoPS is like the idea of "thread" in normal forum, am I right?

**Comment by TheSil** [comment]
No problem, looking forward to see this working! 

> So topic in AoPS is like the idea of "thread" in normal forum, am I right?

Yes, that is correct.

**Comment by t-k-** [comment]
@TheSil  Hi, I have setup a multi-source instance Approach0, I have posted a few message on MSE chat room and you can check out the link there.
